core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt

```diff
@@ -278,23 +276,113 @@ val allLoadingGaugeTypeSet = enumValues<RJSLoadingGaugeType>().toSet()
 private fun parseRjsRouteWaypoint(
     rjsWaypoint: RJSRouteWaypoint,
     builder: RawInfraFromRjsBuilderImpl,
-    trackSectionNameToChunkEndOffset: MutableMap<String, TreeSet<Offset<TrackSection>>>,
+    trackSectionNameToDistanceSortedWaypointIdxs:
+        MutableMap<String, TreeMap<Offset<TrackSection>, DetectorId>>,
     detectorMap: MutableMap<String, DetectorId>,
 ) {
-    detectorMap[rjsWaypoint.id] = builder.detector(rjsWaypoint.id)
+    val waypointIdx = builder.detector(rjsWaypoint.id)
+    detectorMap[rjsWaypoint.id] = waypointIdx
 
-    // Storing only chunk's end: ignore 0 offset that is a chunk's start
-    if (rjsWaypoint.position == 0.0) return
+    if (!trackSectionNameToDistanceSortedWaypointIdxs.containsKey(rjsWaypoint.track)) {
+        trackSectionNameToDistanceSortedWaypointIdxs[rjsWaypoint.track] =
+            TreeMap<Offset<TrackSection>, DetectorId>()
+    }
+    trackSectionNameToDistanceSortedWaypointIdxs[rjsWaypoint.track]!![
+        Offset(rjsWaypoint.position.meters)] = waypointIdx
+}
 
-    if (!trackSectionNameToChunkEndOffset.containsKey(rjsWaypoint.track)) {
-        trackSectionNameToChunkEndOffset[rjsWaypoint.track] = TreeSet<Offset<TrackSection>>()
+/** Build detection zones on a partially-filled builder (track-sections, nodes and detectors) */
+private fun buildZones(
+    builder: RawInfraFromRjsBuilderImpl,
+    trackSectionNameToDistanceSortedWaypointIdxs:
+        Map<String, TreeMap<Offset<TrackSection>, DetectorId>>
+) {
+    // run a union-find to associate track endpoints:
+    // union by node
+    val trackSectionPool = builder.getTrackSectionPool()
+    val trackNodePool = builder.getTrackNodePool()
+    val uf = UnionFind(trackSectionPool.size.toInt() * 2)
+    for (nodeIdx in trackNodePool) {
+        val ports = trackNodePool[nodeIdx].ports
+        var firstEndpoint: EndpointTrackSectionId? = null
+        for (port in ports) {
+            val endpoint = ports[port]
+            if (firstEndpoint == null) {
+                firstEndpoint = endpoint
+            } else {
+                uf.union(firstEndpoint.index.toInt(), endpoint.index.toInt())
+            }
+        }
+    }
+    // union by track-section without waypoint
+    for (trackSectionIdx in trackSectionPool) {
+        val waypoints =
+            trackSectionNameToDistanceSortedWaypointIdxs[trackSectionPool[trackSectionIdx].name]
+        if (waypoints.isNullOrEmpty()) {
+            val startEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.START)
+            val endEndpoint = EndpointTrackSectionId(trackSectionIdx, Endpoint.END)
+            uf.union(startEndpoint.index.toInt(), endEndpoint.index.toInt())
+        }
+    }
+
+    // now that union-find is complete, create detection zones
+    val zoneRootToNodesAndDirDetectors =
+        mutableMapOf<Int, Pair<MutableSet<TrackNodeId>, MutableSet<DirDetectorId>>>()
```

Comment:
The pair of track node and dirdetectorid looks a lot like a the internal zone descriptor. I believe the code would look nicer if the builder API allowed adding track nodes to zones. The track node builder API would then be as follows:

```kotlin
fun zone(): ZoneId
fun zoneAddNode(ZoneId, TrackNodeId)
fun zoneAddBound(ZoneId, DirDetectorId)
```

In this function, you can then do:
```suggestion
val rootToZoneMap = mutableMapOf<Int, ZoneId>()
```

Reply:
Done in https://github.com/osrd-project/osrd/pull/6921/commits/754865f7833fe40281a8789e8f79108f55b6a2dc
Which also triggers https://github.com/osrd-project/osrd/pull/6921/commits/997b08551d1ee2d8032d9d3f279ccde5b11daf78 to allow slightly different result for zones.